server/svix-server/src/v1/endpoints/endpoint/crud.rs

```diff
@@ -125,36 +125,60 @@ pub(super) async fn update_endpoint(
     Path((_app_id, endp_id)): Path<(ApplicationIdOrUid, EndpointIdOrUid)>,
     ValidatedJson(data): ValidatedJson<EndpointIn>,
     AuthenticatedApplication { permissions, app }: AuthenticatedApplication,
-) -> Result<Json<EndpointOut>> {
+) -> Result<(StatusCode, Json<EndpointOut>)> {
     let endp = endpoint::Entity::secure_find_by_id_or_uid(app.id.clone(), endp_id)
         .one(db)
-        .await?
-        .ok_or_else(|| HttpError::not_found(None, None))?;
+        .await?;
 
     if let Some(ref event_types_ids) = data.event_types_ids {
         validate_event_types(db, event_types_ids, &permissions.org_id).await?;
     }
     validate_endpoint_url(&data.url, cfg.endpoint_https_only)?;
 
-    let mut endp: endpoint::ActiveModel = endp.into();
-    data.update_model(&mut endp);
+    let app_uid = app.uid;
 
-    let ret = endp.update(db).await?;
+    let (status_code, ret) = match endp {
+        Some(endp) => {
+            let mut endp: endpoint::ActiveModel = endp.into();
+            data.update_model(&mut endp);
+            let ret = endp.update(db).await?;
+
+            (StatusCode::OK, ret)
+        }
+        None => {
+            let ret = endpoint::ActiveModel {
+                app_id: Set(app.id),
+                key: Set(generate_secret(
+                    &cfg.encryption,
+                    &cfg.default_signature_type,
+                )?),
+                ..data.into()
+            }
+            .insert(db)
+            .await?;
+
+            (StatusCode::CREATED, ret)
+        }
+    };
+
+    let op_event = EndpointEvent {
+        app_id: &ret.app_id,
+        app_uid: app_uid.as_ref(),
+        endpoint_id: &ret.id,
+        endpoint_uid: ret.uid.as_ref(),
+    };
+
+    let op_webhook = match status_code {
+        StatusCode::OK => OperationalWebhook::EndpointUpdated(op_event),
+        StatusCode::CREATED => OperationalWebhook::EndpointCreated(op_event),
```

Comment:
This is way too implicit, it's using external values to this function (the returned state) as internal control flow. The fact that this not the right way is exposed by the `panic!` below. You wouldn't need to panic.

You can just move the operation webhook above (yeah, duplicate for each, it's a minor duplication anyway, and they are inherently different things that just happen to look similar now but maybe change (e.g. we may add "changed fields" to the changed event), so de-duplication them doesn't reduce complexity, but rather adds some.